packages/optimistic-oracle/test/keeper.js

```diff
@@ -0,0 +1,305 @@
+
+    collateralWhitelist = await AddressWhitelist.new();
+    await finder.changeImplementationAddress(utf8ToHex(interfaceName.CollateralWhitelist), collateralWhitelist.address);
+
+    store = await Store.new({ rawValue: "0" }, { rawValue: "0" }, timer.address);
+    await finder.changeImplementationAddress(utf8ToHex(interfaceName.Store), store.address);
+
+    mockOracle = await MockOracle.new(finder.address, timer.address);
+    await finder.changeImplementationAddress(utf8ToHex(interfaceName.Oracle), mockOracle.address);
+  });
+
+  beforeEach(async function() {
+    // Create and save a new collateral token for each request so we can test
+    // that keeper can use different currencies for each request to post bonds.
+    collateralCurrenciesForIdentifier = [];
+    for (let i = 0; i < identifiersToTest.length; i++) {
+      let collateral = await Token.new("Wrapped Ether", "WETH", getPrecisionForIdentifier(identifiersToTest[i]));
+      await collateral.addMember(1, owner);
+      await collateral.mint(botRunner, initialUserBalance);
+      await collateral.mint(requester, initialUserBalance);
+      await collateral.mint(disputer, initialUserBalance);
+      await collateralWhitelist.addToWhitelist(collateral.address);
+      await store.setFinalFee(collateral.address, { rawValue: finalFee });
+      collateralCurrenciesForIdentifier[i] = collateral;
+    }
+
+    optimisticOracle = await OptimisticOracle.new(liveness, finder.address, timer.address);
+
+    // Contract used to make price requests
+    optimisticRequester = await OptimisticRequesterTest.new(optimisticOracle.address);
+
+    startTime = (await optimisticOracle.getCurrentTime()).toNumber();
+    requestTime = startTime - 10;
+
+    spy = sinon.spy();
+    spyLogger = winston.createLogger({
+      level: "info",
+      transports: [new SpyTransport({ level: "info" }, { spy: spy })]
+    });
+
+    client = new OptimisticOracleClient(
+      spyLogger,
+      OptimisticOracle.abi,
+      MockOracle.abi,
+      web3,
+      optimisticOracle.address,
+      mockOracle.address
+    );
+
+    gasEstimator = new GasEstimator(spyLogger);
+  });
+
+  describe("Valid price identifiers", function() {
+    beforeEach(async function() {
+      // Make a new price request for each identifier, each of which should cause the keeper bot to
+      // construct a pricefeed with a new precision.
+      for (let i = 0; i < identifiersToTest.length; i++) {
+        await optimisticRequester.requestPrice(
+          identifiersToTest[i],
+          requestTime,
+          "0x",
+          collateralCurrenciesForIdentifier[i].address,
+          0
+        );
+      }
+
+      // Construct OO Keeper using a valid default price feed config containing any additional properties
+      // not set in DefaultPriceFeedConfig
+      let defaultPriceFeedConfig = {
+        currentPrice: "1", // Mocked current price. This will be scaled to the identifier's precision.
```

Comment:
Ideally, I don't think scaling should occur like this and rather we should define a set of `convertPrice` methods that input defined parameters into the feeds see here within the liquidator for example: https://github.com/UMAprotocol/protocol/blob/c79f3476fee07736257582f7d92eb7c95932c300/packages/liquidator/test/Liquidator.js#L126

If we are to follow this pattern then we should test this bot against a set of different decimals to ensure it can correctly deal with all configurations and scalling. 

LMKWYT.

Reply:
For a set of price requests: A, B, C, how would you determine how to `convertPrice` when responding to each request?

For example, imagine that `A` calls for 6 precision, `B` for 8, and `C` for 18.

Currently, I read from the `DefaultPriceFeed.js` to determine the mapping of identifier --> precision. All I do is call the constructed pricefeed (again, implied from `DefaultPriceFeed`) `getHistoricalPrice()` method which returns price in the identifier's specified precision.

What about this flow can be improved?

Reply:
ye, I see the issue. given that you are generating feeds within your proposer that makes sense to me. In that case I think it's fine.